### PR TITLE
 AWS AppMesh resource changes based on 2019/01/25 API version

### PR DESCRIPTION
**Pull request by ewbankkit**
Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/7659.

### CODE REVIEW

**Comment by ewbankkit** [comment]
Just addresses the changes described [here](https://github.com/awslabs/aws-app-mesh-examples/issues/92).

File: aws/resource_aws_appmesh_mesh.go [hunk 1]
```diff
@@ -71,12 +71,12 @@ func resourceAwsAppmeshMeshRead(d *schema.ResourceData, meta interface{}) error
 	resp, err := conn.DescribeMesh(&appmesh.DescribeMeshInput{
 		MeshName: aws.String(d.Id()),
 	})
+	if isAWSErr(err, "NotFoundException", "") {
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_mesh.go]: Simplified form of error code test.

File: aws/resource_aws_appmesh_route.go [hunk 2]
```diff
@@ -177,8 +181,12 @@ func resourceAwsAppmeshRouteRead(d *schema.ResourceData, meta interface{}) error
 	d.Set("arn", resp.Route.Metadata.Arn)
 	d.Set("created_date", resp.Route.Metadata.CreatedAt.Format(time.RFC3339))
 	d.Set("last_updated_date", resp.Route.Metadata.LastUpdatedAt.Format(time.RFC3339))
-	err1 := d.Set("spec", flattenAppmeshRouteSpec(resp.Route.Spec))
-	return err1
+	err = d.Set("spec", flattenAppmeshRouteSpec(resp.Route.Spec))
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_route.go]: Check and add context to any error.

File: aws/resource_aws_appmesh_mesh.go [hunk 3]
```diff
@@ -100,10 +100,10 @@ func resourceAwsAppmeshMeshDelete(d *schema.ResourceData, meta interface{}) erro
 	_, err := conn.DeleteMesh(&appmesh.DeleteMeshInput{
 		MeshName: aws.String(d.Id()),
 	})
+	if isAWSErr(err, "NotFoundException", "") {
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_mesh.go]: Simplified form of error code test.

File: aws/resource_aws_appmesh_route.go [hunk 4]
```diff
@@ -20,6 +21,9 @@ func resourceAwsAppmeshRoute() *schema.Resource {
 		Read:   resourceAwsAppmeshRouteRead,
 		Update: resourceAwsAppmeshRouteUpdate,
 		Delete: resourceAwsAppmeshRouteDelete,
+		Importer: &schema.ResourceImporter{
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_route.go]: Add ability to `terraform import`.

File: aws/resource_aws_appmesh_route.go [hunk 5]
```diff
@@ -157,12 +161,12 @@ func resourceAwsAppmeshRouteRead(d *schema.ResourceData, meta interface{}) error
 		RouteName:         aws.String(d.Get("name").(string)),
 		VirtualRouterName: aws.String(d.Get("virtual_router_name").(string)),
 	})
+	if isAWSErr(err, appmesh.ErrCodeNotFoundException, "") {
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_route.go]: Simplified form of error code test.

File: aws/resource_aws_appmesh_route.go [hunk 6]
```diff
@@ -212,16 +220,46 @@ func resourceAwsAppmeshRouteDelete(d *schema.ResourceData, meta interface{}) err
...
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_route.go]: Simplified form of error code test.

File: aws/resource_aws_appmesh_mesh_test.go [hunk 7]
```diff
@@ -108,10 +108,10 @@ func testAccCheckAppmeshMeshDestroy(s *terraform.State) error {
 		_, err := conn.DescribeMesh(&appmesh.DescribeMeshInput{
 			MeshName: aws.String(rs.Primary.Attributes["name"]),
 		})
+		if isAWSErr(err, "NotFoundException", "") {
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_mesh_test.go]: Ensure ALL relevant resources have been deleted.

File: aws/resource_aws_appmesh_route_test.go [hunk 8]
```diff
@@ -133,6 +133,12 @@ func testAccAwsAppmeshRoute_basic(t *testing.T) {
 						resourceName, "arn", regexp.MustCompile(fmt.Sprintf("^arn:[^:]+:appmesh:[^:]+:\\d{12}:mesh/%s/virtualRouter/%s/route/%s", meshName, vrName, rName))),
 				),
 			},
+			{
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_route_test.go]: Test import.

File: aws/resource_aws_appmesh_virtual_node.go [hunk 9]
```diff
@@ -44,9 +51,36 @@ func resourceAwsAppmeshVirtualNode() *schema.Resource {
 					Schema: map[string]*schema.Schema{
 						"backends": {
 							Type:     schema.TypeSet,
+							Removed:  "Use `backend` configuration blocks instead",
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_virtual_node.go]: Not going through `Deprecated` cycle here, straight to `Removed` as AppMesh is a preview service.

File: aws/resource_aws_appmesh_virtual_node.go [hunk 10]
```diff
@@ -155,6 +189,13 @@ func resourceAwsAppmeshVirtualNode() *schema.Resource {
 										Elem: &schema.Resource{
 											Schema: map[string]*schema.Schema{
 												"service_name": {
+													Type:     schema.TypeString,
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_virtual_node.go]: Not going through `Deprecated` cycle here, straight to `Removed` as AppMesh is a preview service.

File: aws/resource_aws_appmesh_virtual_router.go [hunk 11]
```diff
@@ -35,18 +41,16 @@ func resourceAwsAppmeshVirtualRouter() *schema.Resource {
 
 			"spec": {
 				Type:     schema.TypeList,
-				Required: true,
-				MinItems: 1,
-				MaxItems: 1,
+				Removed:  "Remove `spec` configuration block completely",
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_virtual_router.go]: Not going through `Deprecated` cycle here, straight to `Removed` as AppMesh is a preview service.

File: aws/resource_aws_appmesh_virtual_router.go [hunk 12]
```diff
@@ -76,7 +80,7 @@ func resourceAwsAppmeshVirtualRouterCreate(d *schema.ResourceData, meta interfac
 	req := &appmesh.CreateVirtualRouterInput{
 		MeshName:          aws.String(d.Get("mesh_name").(string)),
 		VirtualRouterName: aws.String(d.Get("name").(string)),
-		Spec:              expandAppmeshVirtualRouterSpec(d.Get("spec").([]interface{})),
+		Spec:              &appmesh.VirtualRouterSpec{},
```
Comment (ewbankkit) [inline aws/resource_aws_appmesh_virtual_router.go]: `Spec` still needs to be specified here even though it can only be empty.

**Comment by ewbankkit** [comment]
Changes:
* Add `aws_appmesh_virtual_service` resource
* Add ability to import `aws_appmesh_virtual_node`,  `aws_appmesh_virtual_router` and `aws_appmesh_route` resources
* Mark `aws_appmesh_virtual_router.spec` as _Removed_
* Mark `aws_appmesh_virtual_node.backends` as _Removed_, to be replaced by 0 or more `aws_appmesh_virtual_node.backend` configuration blocks
* Mark `aws_appmesh_virtual_node.service_discovery.dns.service_name` as _Removed_, to be replaced by `aws_appmesh_virtual_node.service_discovery.dns.hostname`

**Comment by ewbankkit** [comment]
TODO: Add listeners to `aws_appmesh_virtual_router` resource. See https://github.com/awslabs/aws-app-mesh-examples/issues/94.

**Comment by ewbankkit** [comment]
Incorporates commit from https://github.com/terraform-providers/terraform-provider-aws/pull/7860.

File: aws/resource_aws_appmesh_virtual_router.go (hunk 12 above)
Comment (bcelenza) [reply aws/resource_aws_appmesh_virtual_router.go]: Just a note that this is where the new VirtualRouter `listeners` change goes for https://github.com/awslabs/aws-app-mesh-examples/issues/94, which is a required field in the new APIs.

File: aws/resource_aws_appmesh_virtual_router.go (hunk 11 above)
Comment (bcelenza) [reply aws/resource_aws_appmesh_virtual_router.go]: I assume this message is surfaced to the user? Is there a way to detect values in `service_names` and instruct them to create a VirtualService resource by that name?

File: aws/resource_aws_appmesh_virtual_service.go [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+			},
+
+			"spec": {
+				Type:     schema.TypeList,
+				Required: true,
+				MinItems: 1,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"provider": {
```
Comment (bcelenza) [inline aws/resource_aws_appmesh_virtual_service.go]: Provider in the SDKs is a `union` type, which in AWS' language means only one type can be provided (in this case, either VirtualNode OR VirtualRouter). Just curious how this plays with Terraform? Is it validated here or do you rely on the App Mesh control plane to invalidate when values are provided for both?

Comment (ewbankkit) [reply aws/resource_aws_appmesh_virtual_service.go]: * RE https://github.com/terraform-providers/terraform-provider-aws/pull/7858#discussion_r263634391: Once I add listeners then I can move `Removed` down to `service_names` and change the message to mention the virtual service resource.

* RE https://github.com/terraform-providers/terraform-provider-aws/pull/7858#discussion_r263635145: Yes, we can do that with a `ConflictsWith` attribute

**Comment by ewbankkit** [comment]
This is ready for review but I'm not seeing the button to move out of Draft status.
High level changes are listed [above](https://github.com/terraform-providers/terraform-provider-aws/pull/7858#issuecomment-470705476).
I have added documentation to the `aws_appmesh_virtual_node` and `aws_appmesh_virtual_resources` describing the code changes required.
Terraform state for existing resources is being migrated.

**Comment by ewbankkit** [comment]
Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAWSAppmesh'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAWSAppmesh -timeout 120m
=== RUN   TestAccAWSAppmesh
=== RUN   TestAccAWSAppmesh/Mesh
=== RUN   TestAccAWSAppmesh/Mesh/basic
=== RUN   TestAccAWSAppmesh/Route
=== RUN   TestAccAWSAppmesh/Route/basic
=== RUN   TestAccAWSAppmesh/Route/allAttributes
=== RUN   TestAccAWSAppmesh/VirtualNode
=== RUN   TestAccAWSAppmesh/VirtualNode/basic
=== RUN   TestAccAWSAppmesh/VirtualNode/allAttributes
=== RUN   TestAccAWSAppmesh/VirtualRouter
=== RUN   TestAccAWSAppmesh/VirtualRouter/basic
=== RUN   TestAccAWSAppmesh/VirtualService
=== RUN   TestAccAWSAppmesh/VirtualService/virtualNode
=== RUN   TestAccAWSAppmesh/VirtualService/virtualRouter
--- PASS: TestAccAWSAppmesh (186.72s)
    --- PASS: TestAccAWSAppmesh/Mesh (15.21s)
        --- PASS: TestAccAWSAppmesh/Mesh/basic (15.21s)
    --- PASS: TestAccAWSAppmesh/Route (44.83s)
        --- PASS: TestAccAWSAppmesh/Route/basic (17.80s)
        --- PASS: TestAccAWSAppmesh/Route/allAttributes (27.03s)
    --- PASS: TestAccAWSAppmesh/VirtualNode (41.86s)
        --- PASS: TestAccAWSAppmesh/VirtualNode/basic (16.59s)
        --- PASS: TestAccAWSAppmesh/VirtualNode/allAttributes (25.27s)
    --- PASS: TestAccAWSAppmesh/VirtualRouter (26.98s)
        --- PASS: TestAccAWSAppmesh/VirtualRouter/basic (26.98s)
    --- PASS: TestAccAWSAppmesh/VirtualService (57.82s)
        --- PASS: TestAccAWSAppmesh/VirtualService/virtualNode (30.45s)
        --- PASS: TestAccAWSAppmesh/VirtualService/virtualRouter (27.37s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	186.740s
```

**Comment by bflad** [comment]
> This is ready for review but I'm not seeing the button to move out of Draft status.

Interesting -- I was able to see and click a "Ready for review" button near the "Merge pull request" button. Clicked it for you.

**Comment by ewbankkit** [comment]
@bflad Thanks.

**Comment by ewbankkit** [comment]
Rebased to fix merge conflict on `go.sum`.

**Comment by bflad** [comment]
This has been released in [version 2.3.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#230-march-21-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**